share/shwap/range_namespace_data_id.go

```diff
@@ -0,0 +1,158 @@
+func (rngid RangeNamespaceDataID) Validate() error {
+	if err := rngid.DataNamespace.ValidateForData(); err != nil {
+		return fmt.Errorf("RangeNamespaceDataID: namespace validation: %w", err)
+	}
+	return rngid.SampleID.Validate()
+}
+
+// ReadFrom reads the binary form of RangeNamespaceDataID from the provided reader.
+func (rngid *RangeNamespaceDataID) ReadFrom(r io.Reader) (int64, error) {
+	data := make([]byte, RangeNamespaceDataIDSize)
+	n, err := io.ReadFull(r, data)
+	if err != nil {
+		return int64(n), err
+	}
+	if n != RangeNamespaceDataIDSize {
+		return int64(n), fmt.Errorf("RangeNamespaceDataID: expected %d bytes, got %d", RangeNamespaceDataIDSize, n)
+	}
+
+	id, err := RangeNamespaceDataIDFromBinary(data)
+	if err != nil {
+		return int64(n), fmt.Errorf("RangeNamespaceDataIDFromBinary: %w", err)
+	}
+	*rngid = id
+	return int64(n), nil
+}
+
+// WriteTo writes the binary form of RangeNamespaceDataID to the provided writer.
+func (rngid RangeNamespaceDataID) WriteTo(w io.Writer) (int64, error) {
+	data, err := rngid.MarshalBinary()
+	if err != nil {
+		return 0, err
+	}
+	n, err := w.Write(data)
+	return int64(n), err
+}
+
+// Equals checks equality of RangeNamespaceDataID.
+func (rngid *RangeNamespaceDataID) Equals(other RangeNamespaceDataID) bool {
+	return rngid.SampleID.Equals(other.SampleID) && rngid.DataNamespace.Equals(other.DataNamespace) &&
+		rngid.To == other.To
+}
+
+// RangeNamespaceDataIDFromBinary deserializes a RangeNamespaceDataID from its binary form.
+func RangeNamespaceDataIDFromBinary(data []byte) (RangeNamespaceDataID, error) {
+	if len(data) != RangeNamespaceDataIDSize {
+		return RangeNamespaceDataID{}, fmt.Errorf(
+			"invalid RangeNamespaceDataID data length: expected %d, got %d", RangeNamespaceDataIDSize, len(data),
+		)
+	}
+
+	sid, err := SampleIDFromBinary(data[:SampleIDSize])
+	if err != nil {
+		return RangeNamespaceDataID{}, fmt.Errorf("converting SampleId from binary: %w", err)
+	}
+
+	ns, err := libshare.NewNamespaceFromBytes(data[SampleIDSize : libshare.NamespaceSize+SampleIDSize])
+	if err != nil {
+		return RangeNamespaceDataID{}, fmt.Errorf("converting namespace from binary: %w", err)
+	}
+
+	toCoords := SampleCoords{
+		Row: int(binary.BigEndian.Uint16(data[libshare.NamespaceSize+SampleIDSize : libshare.NamespaceSize+SampleIDSize+2])),
+		Col: int(binary.BigEndian.Uint16(data[libshare.NamespaceSize+SampleIDSize+2:])),
+	}
+
+	rngID := RangeNamespaceDataID{
+		SampleID:      sid,
+		DataNamespace: ns,
+		To:            toCoords,
+	}
+	return rngID, rngID.Validate()
+}
+
+// MarshalBinary encodes RangeNamespaceDataID into binary form.
+func (rngid RangeNamespaceDataID) MarshalBinary() ([]byte, error) {
+	data := make([]byte, 0, RangeNamespaceDataIDSize)
+	return rngid.appendTo(data), nil
+}
+
+// appendTo helps in constructing the binary representation  of RangeNamespaceDataID
+// by appending all encoded fields.
+func (rngid RangeNamespaceDataID) appendTo(data []byte) []byte {
```

Comment:
Better to use `AppendBinary(b []byte) ([]byte, error)` (which should land in 1.24 https://github.com/golang/go/issues/62384)

Reply:
Done https://github.com/celestiaorg/celestia-node/issues/4079